mesonbuild/compilers/cpp.py

```diff
@@ -216,8 +216,9 @@ def get_options(self) -> 'MutableKeyedOptionDictType':
         })
         opts[key.evolve('std')].choices = [
             'none', 'c++98', 'c++03', 'c++11', 'c++14', 'c++17', 'c++1z',
-            'c++2a', 'c++20', 'gnu++11', 'gnu++14', 'gnu++17', 'gnu++1z',
-            'gnu++2a', 'gnu++20',
+            'c++2a', 'c++20', 'c++2b', 'c++23', 'c++2c', 'c++26', 'gnu++11',
+            'gnu++14', 'gnu++17', 'gnu++1z', 'gnu++2a', 'gnu++20', 'gnu++2b',
+            'gnu++23', 'gnu++2c', 'gnu++26'
```

Comment:
I'd prefer to do these the way e.g. VisualStudioCPPCompiler does it:

https://github.com/mesonbuild/meson/blob/f92bc05c182d9243a19c4af8c303986f1719eee6/mesonbuild/compilers/cpp.py#L733-L743

Reply:
It gets more complicated when you consider that Apple Clang's versioning differs from upstream Clang.

Reply:
```python
class AppleClangCPPCompiler(ClangCPPCompiler):
```

can be overridden, so it is definitely possible somehow... how to make it pretty, hmm...

Reply:
I'm not sure why it isn't simply a runtime check anyway, at least with Clang/GCC. You can do `cc -std=blah -c /dev/null -o /dev/null` and it'll have a nonzero exit code if you provide a standard option that isn't supported. Why the hardcoded list of acceptable standard names?

Reply:
We can follow the guidance of compilers/c.py here:
https://github.com/mesonbuild/meson/blob/f92bc05c182d9243a19c4af8c303986f1719eee6/mesonbuild/compilers/c.py#L114-L142
https://github.com/mesonbuild/meson/blob/f92bc05c182d9243a19c4af8c303986f1719eee6/mesonbuild/compilers/c.py#L195-L205

Reply:
Can this refactoring be done in another PR?  This patch seems to follow the current conventions.

As it stands there is no way to use `c++23` with Clang, which is pretty unfortunate, and this commit will resolve it.

Reply:
Per my comparison this is *not* following the current conventions. ;)

Reply:
Can you explain?  

This commit is mostly adding new values to existing arrays.  That seems to be me to be "current conventions".  Maybe it isn't a best practice, and as you pointed out, other compilers might be doing it differently, but it seems to be current convention in the scope of gcc/clang.

Reply:
@williamspatrick it would look something like this for the clang cpp compiler:

```python
        std_opt = opts[key.evolve('std')]
        std_opt.choices = [
            'none', 'c++98', 'c++03', 'c++11', 'c++14', 'c++17', 'c++1z',
            'c++2a', 'c++20', 'gnu++11', 'gnu++14', 'gnu++17', 'gnu++1z',
            'gnu++2a', 'gnu++20',
        ]

        if version_compare(self.version, '>= version that added C++ 23')
            std_opt.choices.extend(['c++23 aliases'])
        # same for C++26
```

Here is the relebant snippet in the GnuCPPCompiler:
```python
        cppstd_choices = [
            'none', 'c++98', 'c++03', 'c++11', 'c++14', 'c++17', 'c++1z',
            'c++2a', 'c++20', 'gnu++03', 'gnu++11', 'gnu++14', 'gnu++17',
            'gnu++1z', 'gnu++2a', 'gnu++20',
        ]
        if version_compare(self.version, '>=12.2.0'):
            cppstd_choices.append('c++23')
            cppstd_choices.append('gnu++23')
```

Reply:
Pretty sure this is what @eli-schwartz is getting at.

Reply:
> Can you explain?
> 
> This commit is mostly adding new values to existing arrays. That seems to be me to be "current conventions". Maybe it isn't a best practice, and as you pointed out, other compilers might be doing it differently, but it seems to be current convention in the scope of gcc/clang.

I lack the excitement to quibble over whether "current conventions" should be graded by the convention for the current function or the convention for the codebase, or whether, if graded by function-local convention, that is a good rationale for violating a best practice.

To me, it's good enough that I submitted a review with requested changes weeks ago, and that review had two suggestions, and neither of those suggestions have been responded to. You're only arguing in favor of overlooking *one* of those review suggestions -- what about the other one?

But, let's not talk about whether function-local conventions or project conventions qualify as conventions. Let's talk about facts. The current PR modifies this:

```python
cppstd_choices = [
    'none', 'c++98', 'c++03', 'c++11', 'c++14', 'c++17', 'c++1z',
    'c++2a', 'c++20', 'gnu++03', 'gnu++11', 'gnu++14', 'gnu++17',
    'gnu++1z', 'gnu++2a', 'gnu++20',
]
if version_compare(self.version, '>=12.2.0'):
    cppstd_choices.append('c++23')
    cppstd_choices.append('gnu++23')
```

to this:
```python
cppstd_choices = [
    'none', 'c++98', 'c++03', 'c++11', 'c++14', 'c++17', 'c++1z',
    'c++2a', 'c++20', 'c++2b', 'c++23', 'c++2c', 'c++26', 'gnu++03',
    'gnu++11', 'gnu++14', 'gnu++17', 'gnu++1z', 'gnu++2a', 'gnu++20',
    'gnu++2b', 'gnu++23', 'gnu++2c', 'gnu++26'
]
if version_compare(self.version, '>=12.2.0'):
    cppstd_choices.append('c++23')
    cppstd_choices.append('gnu++23')
```

This is pretty factually a fact that within a function that already has a version_compare adding a specific std, this PR proposes to add those same stds a second time -- so that they appear twice in the list -- except without the version_compare.

There is absolutely nothing to debate here about best practices or the lack thereof. The GnuCPPCompiler component of this PR, at least, has an outright coding error in it; this needs to be fixed before the PR can be merged independent of whether I accept your argument about current conventions.

If my suggestion had been addressed, then that would have automatically included fixing this coding error. But for the purposes of ending this debate, I guess I just have to spell it out.

Reply:
I'm not going to be available to make changes to this PR for about two weeks. If someone else would like to take over and make whatever changes are needed, I would appreciate it.

The only part actually relevant to my initial request is the change for `objcpp.py`, so please preserve that, whoever takes over...

Reply:
@eli-schwartz - Thank you for the details.  It wasn't obvious to me that you were pointing out a bug.  I must still be missing your "other suggestion" because it seems this comment sequence started with "VisualStudio...".

@tycho - I might try to take over this tomorrow since @tristan957 mentions something would need to be done by Sunday for the next release.  I really only care about getting C++23 working on Clang and I've tested this sequence there.  Since Clang and GCC currently have different Meson support for C++23, which is pretty frustrating, that is honestly my priority and not ObjC.

If someone else is also interested in taking this further, I'm more than willing to step aside as well.